### PR TITLE
AIRCoP softmax

### PR DESCRIPTION
**Pull request by luke-jiang**
Implement AIRCoP softmax and activation

- [x] how to deal with softmax algorithm and mode? Should user be able to specify that?
- [x] `getConvDescriptor` doesn't add back to map
- [x] `finalize_cudnn` doesn't emit typeless DESTROY calls
- [x] create test case for activation
- [x] change `overwriteCheckFiles` to false before merge

### CODE REVIEW

File: src/out/transformer/distributed_tensor/softmax.check.cu [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+    int x52 = 0;
+    while (x52 != 18) {
+      printf("%f ", x51[x52]);
+      x52 = x52 + 1;
+    }
+    printf("\n");
+    // end copying GPU array x298 to CPU and print for size 18 and type Float
+  }
+  printf("compile");
+  CUDNNCHECK(cudnnDestroy(x8));
```
Comment (luke-jiang) [inline src/out/transformer/distributed_tensor/softmax.check.cu:177]: should also include CUDNNCHECK(cudnnDestroyTensorDescriptor(x39)), since x39 is a descriptor created, and should be destroyed by finalize_cudnn

File: src/main/scala/lms/transformation/distributed_tensor/distributedTensorConv.scala [hunk 2]
```diff
@@ -71,6 +71,22 @@ trait FixedSizeDistributedTensorConvTypeLess extends FixedSizeDistributedTensorM
 
     TensorType(output_shape, input.et, anno)
   }
+  
+  def SoftmaxForward(input: TENSOR, params: SoftmaxParam, anno: Anno, __pos: SourceContext): TENSOR = {
+    val SoftmaxParam(alpha, beta) = params
```
Comment (feiwang3311) [inline src/main/scala/lms/transformation/distributed_tensor/distributedTensorConv.scala:76]: this line seems not necessary

File: src/main/scala/lms/transformation/distributed_tensor/distributedTensorConv.scala [hunk 3]
```diff
@@ -71,6 +71,22 @@ trait FixedSizeDistributedTensorConvTypeLess extends FixedSizeDistributedTensorM
...
+
+    val res_tt = input.resultType
+    (new TENSOR(Adapter.g.reflectRead("tensor_softmax", C(res_tt), C(anno), input.x, 
+      C(params))(input.x)).withSrcType(__pos, input.et))
+  }
+
+  def SoftmaxBackward(output: TENSOR, doutput: TENSOR, params: SoftmaxParam, anno: Anno, __pos: SourceContext): TENSOR = {
+    val SoftmaxParam(alpha, beta) = params
```
Comment (feiwang3311) [inline src/main/scala/lms/transformation/distributed_tensor/distributedTensorConv.scala:84]: this line seems not necessary

File: src/main/scala/lms/transformation/distributed_tensor/distributedTensorConv.scala [hunk 4]
```diff
@@ -86,6 +102,8 @@ trait FixedSizeDistributedTensorConvTypeLess extends FixedSizeDistributedTensorM
       val filterCin   = filter_type(CUDNN_C_IN).dim
       List((inputC, filterCin), (outputC, filterCout))
 
+    case Node(s, op, _, _) if op == "tensor_softmax" => List()
```
Comment (feiwang3311) [inline src/main/scala/lms/transformation/distributed_tensor/distributedTensorConv.scala:105]: would be simpler to just do `case Node(s, "tensor_softmax", _, _) => List()`

File: src/main/scala/lms/transformation/distributed_tensor/distributedTensorConv.scala [hunk 5]
```diff
@@ -111,6 +129,16 @@ trait FixedSizeDistributedTensorConvTypeLess extends FixedSizeDistributedTensorM
...
         }) +=: backwardNodes
+      
+      case Node(s, "tensor_softmax", tt::Backend.Const(anno:Anno)::(a:Backend.Sym)::Backend.Const(params:SoftmaxParam)::_, _) =>
+        implicit val pos = Adapter.oldSourceMap(s)
+        forwardNodes += node
+        (() => {
+            val x = new TENSOR(transform(a))
+            val grad = SoftmaxBackward(x, gradMap(s), params, anno, pos)
+            Accumulate(gradMap(a), grad, anno); ()
+           ()
```
Comment (feiwang3311) [inline src/main/scala/lms/transformation/distributed_tensor/distributedTensorConv.scala:140]: I am not sure about this one:
should it be
`val grad = SoftmaxBackward(x, gradMap(s), params, anno, pos)`
or
`val grad = SoftmaxBackward(new TENSOR(transform(s)), gradMap(s), params, anno, pos)`?

The API https://docs.nvidia.com/deeplearning/cudnn/api/index.html#cudnnSoftmaxBackward seems to be using `yDesc` and `y`, but I am not certain.

Comment (luke-jiang) [reply src/main/scala/lms/transformation/distributed_tensor/distributedTensorConv.scala:153]: I think it's the second one, but I have an error when running on the server.

File: src/main/scala/lms/transformation/distributed_tensor/distributedTensorConv.scala (hunk 2 above)
Comment (luke-jiang) [reply src/main/scala/lms/transformation/distributed_tensor/distributedTensorConv.scala]: fixed in 232d098

File: src/main/scala/lms/transformation/distributed_tensor/distributedTensorConv.scala (hunk 3 above)
Comment (luke-jiang) [reply src/main/scala/lms/transformation/distributed_tensor/distributedTensorConv.scala]: fixed in 232d098

File: src/main/scala/lms/transformation/distributed_tensor/distributedTensorConv.scala (hunk 4 above)
Comment (luke-jiang) [reply src/main/scala/lms/transformation/distributed_tensor/distributedTensorConv.scala]: fixed in 232d098

File: src/out/transformer/distributed_tensor/softmax.check.cu [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+  if (x46) {
+    int x51 = x1;
+    int x52 = 0;
+    while (x52 != x51) {
+      int x53 = x52;
+      NCCLCHECK(ncclRecv(x47 + x53 * 9, (size_t)18, ncclFloat32, x53, x5, x6));
+      x52 = x52 + 1;
+    }
+  }
+  NCCLCHECK(ncclGroupEnd());
```
Comment (luke-jiang) [inline src/out/transformer/distributed_tensor/softmax.check.cu:172]: NCCL failure softmax.check.cu:172 'unhandled cuda error'

File: src/test/scala/lms/utils.scala [hunk 7]
```diff
@@ -10,7 +10,7 @@ trait LibSuite extends FunSuite {
 }
 
 trait TutorialFunSuite extends LibSuite {
-  val overwriteCheckFiles = false // should be false; temporary set to true only to simplify development
+  val overwriteCheckFiles = true // should be false; temporary set to true only to simplify development
```
Comment (feiwang3311) [inline src/test/scala/lms/utils.scala:13]: don't forget to revert this change.

**Review by feiwang3311 — APPROVED** [review]
LGTM Thanks!

File: src/test/scala/lms/utils.scala (hunk 7 above)
Comment (luke-jiang) [reply src/test/scala/lms/utils.scala]: fixed in 5deaa7c
